File: csharp/src/Vista.SDK/MetadataTagsQuery.cs
namespace Vista.SDK;

public sealed record MetadataTagsQuery
{
    internal MetadataTagsQueryBuilder Builder { get; }

    internal MetadataTagsQuery(MetadataTagsQueryBuilder builder) => Builder = builder;

    public bool Match(LocalId? localId) => Builder.Match(localId);
}

public sealed record MetadataTagsQueryBuilder
{
    private readonly Dictionary<CodebookName, MetadataTag> _tags = new();

    private bool _matchExact;

    private MetadataTagsQueryBuilder() { }

    public static MetadataTagsQueryBuilder Empty() => new();

    public static MetadataTagsQueryBuilder From(LocalId localId, bool allowOtherTags = true)
    {
        var builder = new MetadataTagsQueryBuilder();
        foreach (var tag in localId.MetadataTags)
            builder = builder.WithTag(tag);
        builder = builder.WithAllowOtherTags(allowOtherTags);
        return builder;
    }

    public MetadataTagsQuery Build() => new(this);

    public MetadataTagsQueryBuilder WithTag(CodebookName name, string value) => WithTag(new(name, value));

    public MetadataTagsQueryBuilder WithTag(MetadataTag tag)
    {
        _tags[tag.Name] = tag;
        return this with { };
    }

    public MetadataTagsQueryBuilder WithAllowOtherTags(bool allowOthers)
    {
        return this with { _matchExact = !allowOthers };
    }

    internal bool Match(LocalId? localId)
    {
        if (localId is null)
            return false;

        var metadataTags = localId.MetadataTags.ToDictionary(t => t.Name);

        if (_tags.Count > 0)
        {
            if (_matchExact)
            {
                return _tags.Count == metadataTags.Count
                    && _tags.All(t => metadataTags.TryGetValue(t.Key, out var value) && t.Value.Equals(value));
            }
            else
            {
                foreach (var tag in _tags.Values)
                {
                    if (!metadataTags.TryGetValue(tag.Name, out var otherTag))
                        return false;
                    if (tag.Equals(otherTag) == false)
                        return false;
                }
                return true;
            }
        }
        else
        {
            return !_matchExact;
        }
    }
}

File: cpp/src/SDK/query/MetadataTagsQuery.cpp
#include "dnv/vista/sdk/query/MetadataTagsQuery.h"

#include "dnv/vista/sdk/core/LocalId.h"

namespace dnv::vista::sdk
{
    MetadataTagsQuery::MetadataTagsQuery(MetadataTagsQueryBuilder builder)
        : m_builder{ std::move(builder) }
    {}

    bool MetadataTagsQuery::match(const LocalId& localId) const
    {
        const auto& tags = m_builder.m_tags;
        const bool matchExact = m_builder.m_matchExact;

        std::unordered_map<CodebookName, MetadataTag> metadataTags;
        for (const auto& tag : localId.metadataTags())
        {
            metadataTags.emplace(tag.name(), tag);
        }

        if (!tags.empty())
        {
            if (matchExact)
            {
                if (tags.size() != metadataTags.size())
                {
                    return false;
                }

                for (const auto& [name, tag] : tags)
                {
                    auto it = metadataTags.find(name);
                    if (it == metadataTags.end() || it->second != tag)
                    {
                        return false;
                    }
                }
                return true;
            }
            else
            {
                // Subset match: must have AT LEAST these tags (can have more)
                for (const auto& [name, tag] : tags)
                {
                    auto it = metadataTags.find(name);
                    if (it == metadataTags.end())
                    {
                        return false;
                    }
                    if (it->second != tag)
                    {
                        return false;
                    }
                }
                return true;
            }
        }
        else
        {
            // No tags specified:
            // - If matchExact (allowOthers=false): match only LocalIds with no tags
            // - If !matchExact (allowOthers=true): match any LocalId
            return !matchExact;
        }
    }

    const MetadataTagsQueryBuilder& MetadataTagsQuery::builder() const noexcept
    {
        return m_builder;
    }

    MetadataTagsQueryBuilder MetadataTagsQueryBuilder::from(const LocalId& localId, bool allowOtherTags)
    {
        MetadataTagsQueryBuilder builder;
        for (const auto& tag : localId.metadataTags())
        {
            builder = std::move(builder).withTag(tag);
        }
        return std::move(builder).withAllowOtherTags(allowOtherTags);
    }
} // namespace dnv::vista::sdk
